clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java

```diff
@@ -263,6 +265,90 @@ protected RegisterTaskDefinitionRequest makeTaskDefinitionRequest(
     return request;
   }
 
+  private RegisterTaskDefinitionRequest makeTaskDefinitionRequestFromArtifact(
+      String ecsServiceRole, String newServerGroupName) {
+
+    File artifactFile =
+        downloadTaskDefinitionArtifact(description.getResolvedTaskDefinitionArtifact());
+
+    RegisterTaskDefinitionRequest requestTemplate;
+    try {
+      requestTemplate = mapper.readValue(artifactFile, RegisterTaskDefinitionRequest.class);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    List<ContainerDefinition> containers = requestTemplate.getContainerDefinitions();
+    if (containers.size() == 0) {
+      throw new IllegalArgumentException(
+          "Provided task definition does not contain any container definitions.");
+    }
+
+    description
+        .getContainerToImageMap()
+        .forEach(
+            (k, v) -> {
+              // check if taskDefTemplate contains matching container
+              List<ContainerDefinition> matches =
+                  containers.stream()
+                      .filter(x -> x.getName().equals(k))
+                      .collect(Collectors.toList());
+
+              if (matches.size() != 1) {
+                throw new IllegalArgumentException(
+                    "Invalid number of matching containers found for mapping '"
+                        + k
+                        + "'. Have "
+                        + matches.size()
+                        + " but expected 1.");
+              }
+
+              // interpolate container mappings
+              matches.get(0).setImage(v);
+            });
+
+    containers.forEach(
+        (c) -> {
+          Collection<KeyValuePair> updatedEnv =
+              setSpinnakerEnvVars(c.getEnvironment(), newServerGroupName);
+          c.setEnvironment(updatedEnv);
+
+          Map<String, String> updatedLabels =
+              setSpinnakerDockerLabels(c.getDockerLabels(), newServerGroupName);
+          c.setDockerLabels(updatedLabels);
+        });
+
+    if (FARGATE_LAUNCH_TYPE.equals(description.getLaunchType())) {
+      requestTemplate.setExecutionRoleArn(ecsServiceRole);
```

Comment:
I think ideally if the role is already specified in the template, we should leave it there, and only use the Spinnaker role if it's missing.
Related: https://github.com/spinnaker/spinnaker/issues/3340